File: dynamic-array.go
package main

import "fmt"

type DynamicArray struct {
	Capacity int
	Length   int
	Arr      []int
}

func NewDynamicArray() *DynamicArray {
	return &DynamicArray{
		Capacity: 2,
		Length:   0,
		Arr:      make([]int, 2),
	}
}

func (d *DynamicArray) push(n int) {
	//If length has reached capacity then resize
	if d.Length == d.Capacity {
		d.resize()
	}

	//Push to end
	d.Arr[d.Length] = n
	d.Length++
}

func (d *DynamicArray) resize() {
	d.Capacity = d.Capacity * 2
	newArr := make([]int, d.Capacity)

	//can also use copy(newArr, d.Arr)
	for i := 0; i < len(d.Arr); i++ {
		newArr[i] = d.Arr[i]
	}

	d.Arr = newArr
}

func (d *DynamicArray) pop() {
	if d.Length > 0 {
		d.Arr[d.Length] = 0
		d.Length--
	}
}

func (d *DynamicArray) Get(i int) int {
	if i < d.Length {
		return d.Arr[i]
	}
	return -1
}

func (d *DynamicArray) Insert(i, val int) {
	if i < d.Length {
		d.Arr[i] = val
	}
}

func (d *DynamicArray) Print() {
	for i := 0; i < d.Length; i++ {
		fmt.Println(d.Arr[i])
	}
}

File: DynamicArray.java
public class DynamicArray {
    int capacity;
    int length;
    int[] array;

    public DynamicArray() {
        capacity = 2;
        length = 0;
        array = new int[capacity];
    }

    /**
     * Insert at end - AKA - push
     */
    public void push(int value) {
        if(length == capacity) {
            this.resize();
        }

        array[length] = value;
        length++;
    }

    /**
     * Resize the array to double the size it was before
     * Resizing involves creating a new array and then copying
     * the elements from old to new array
     */
    public void resize() {
        capacity = capacity * 2;
        int[] newArray = new int[capacity];

        for(int i = 0; i < array.length; i++) {
            newArray[i] = array[i];
        }
        array = newArray;
    }

    /**
     * Remove last element
     * Since we use the length property to access the last element
     * decrementing the length by 1 will make sure that element is never accessed
     */
    public void pop() {
        if(length > 0) {
            length--;
        }
    }

    public int getAtIndex(int index) {
        if(index < length) {
            return array[index];
        }

        return -1;
    }

    public void insertAtIndex(int index, int value) {
        if(index < length) {
            array[index] = value;
            return;
        }

        return; // or throw exception
    }
}
